lib/much-rails/action.rb

```diff
@@ -0,0 +1,394 @@
+    attr_reader :on_call_block, :on_before_call_blocks, :on_after_call_blocks
+
+    def initialize
+      @params_root = nil
+      @required_params = []
+      @date_params = []
+      @time_params = []
+      @boolean_params = []
+
+      @format = nil
+
+      @on_validation_blocks = []
+      @on_before_call_blocks = []
+      @on_after_call_blocks = []
+      @on_call_block = -> {}
+    end
+
+    def set_params_root(value)
+      @params_root = value
+    end
+
+    def add_required_params(param_names)
+      @required_params.concat(Array.wrap(param_names))
+    end
+
+    def add_date_params(klass, param_names)
+      Array.wrap(param_names).tap { |names|
+        @date_params.concat(names)
+
+        names.each do |name|
+          klass.public_send(:define_method, name, &memoize_date_decode(name))
+        end
+      }
+    end
+
+    def add_time_params(klass, param_names)
+      Array.wrap(param_names).tap { |names|
+        @time_params.concat(names)
+
+        names.each do |name|
+          klass.public_send(:define_method, name, &memoize_time_decode(name))
+        end
+      }
+    end
+
+    def add_boolean_params(klass, param_names)
+      Array.wrap(param_names).tap { |names|
+        @boolean_params.concat(names)
+
+        names.each do |name|
+          klass.public_send(
+            :define_method, "#{name}?", &memoize_boolean_decode(name))
+        end
+      }
+    end
+
+    def set_format(value)
+      @format = value
+    end
+
+    def add_on_validation_block(block)
+      @on_validation_blocks << block
+    end
+
+    def set_on_call_block(block)
+      @on_call_block = block
+    end
+
+    def add_on_before_call_block(block)
+      @on_before_call_blocks << block
+    end
+
+    def add_on_after_call_block(block)
+      @on_after_call_blocks << block
+    end
+
+    private
+
+    def memoize_date_decode(method_name)
+      -> {
+        instance_variable_name = "@#{method_name}"
+        if instance_variable_get(instance_variable_name).nil?
+          instance_variable_set(
+            instance_variable_name,
+            MuchRails::Date.for(params[method_name])
+          )
+        end
+        instance_variable_get(instance_variable_name)
+      }
+    end
+
+    def memoize_time_decode(method_name)
+      -> {
+        instance_variable_name = "@#{method_name}"
+        if instance_variable_get(instance_variable_name).nil?
+          instance_variable_set(
+            instance_variable_name,
+            MuchRails::Time.for(params[method_name])
+          )
+        end
+        instance_variable_get(instance_variable_name)
+      }
+    end
+
+    def memoize_boolean_decode(method_name)
+      -> {
+        instance_variable_name = "@#{method_name}"
+        if instance_variable_get(instance_variable_name).nil?
+          instance_variable_set(
+            instance_variable_name,
+            MuchRails::Boolean::FALSE_VALUES.exclude?(params[method_name])
```

Comment:
Does this work with `MuchRails::Boolean` (aka MuchBoolean)? I don't think this is the API to convert boolean values. I think it is `MuchRails::Boolean.convert(params[method_name])`. See https://github.com/redding/much-boolean.

Reply:
It works, but `MuchRails::Boolean.convert(params[method_name])` is a cleaner, less-coupled approach.